platform/BufferedSerial.h

```diff
@@ -0,0 +1,94 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef MBED_BUFFEREDSERIAL_H
+#define MBED_BUFFEREDSERIAL_H
+
+#include "platform/platform.h"
+
+#if DEVICE_SERIAL
+
+#include "FileHandle.h"
+#include "SerialBase.h"
+#include "PlatformMutex.h"
+#include "serial_api.h"
+#include "CircularBuffer.h"
+
+namespace mbed {
+class BufferedSerial : private SerialBase, public FileHandle {
```

Comment:
This isn't a pattern we've followed with other external libraries (EventQueue comes to mind). Why the change? Changing a class name (a code change) is more disruptive than removing a dependency. This is just going to add another barrier to mainlining external projects in the future.

The ATParser and BufferedSerial modifications can be backwards compatible. Should the changes be first merged into the external repos? Would that help with migration?

> The API changes are fundamental to making the ATParser generic and not tied to the serial port.

Is this immediately required? So far every instance of ATParser I can find is using BufferedSerial (Including Rob's cellular implementation). It's a nice feature to have but seems inappropriate if it's blocking the cellular branch from getting into 5.5.